packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+# Nimble Table Columns API
+
+## Problem Statement
+
+The `nimble-table` requires users to be able to configure which columns to display for a table. Additionally, a particular display column may require access to multiple fields in the data, and allow a user to either define the field to sort by, or provide a custom sort routine. We need to provide a means for a client to provide their own column visualization. Finally, columns must also provide the user to either specify the header text, or custom visuals to use for the header.
+
+### Out of scope of this HLD
+
+Programmatic API for state that could be considered column-centric: width, sort direction, grouped, etc. These concerns should be covered in separate designs covering those topics specifically, allowing for discussion on both the interactive side and the API design on an individual basis.
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```
+<nimble-table>
+    <nimble-table-column-text valueKey="name"></nimble-table-column-text>
+    <nimble-table-column-icon valueKey="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation.
+
+The table API to support this could look like the following:
+
+```TS
+public Table extends FoundationElement {
+    ...
+
+    @observable
+    public slottedColumns: ITableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `ITableColumn`
+
+This interface is what a column web component (i.e. a slotted column) must implement.
+
+```TS
+interface ITableColumn {
+    header?: string; // the string to use in the header for the column
+    headerTemplateFn?: () => ViewTemplate; // the template to use for the header (string if none provided)
+    cellTemplateFn: () => ViewTemplate<ITableCellData>; // the template to use to render the cell content for the column (required)
```

Comment:
why a headerTemplate function instead of a headerTemplate? The examples shown below for cellTemplateFn don't actually return a function that returns a template.

I'm also curious why not a readonly property instead? I'm hoping and maybe it should be spelled out in the spec that the template itself will only every be made once. The template instance can be re-used and bound many times in many instances but the template should only be created once.

See the note at the end of [composing templates](https://www.fast.design/docs/fast-element/using-directives/#composing-templates):

> When composing templates, extract the composed template to an external variable. If you define the template inline, within your method, property, or expression, then each time that is invoked, a new instance of the template will be created, rather than reusing the template. This will result in an unnecessary performance cost.

So yea I'd consider a property a better API as it emphasizes that we are only going to access a single instance and only one instance should be made (but that template will be re-used and bound in many cells).